src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/arrival/ArrivalsService.kt

```diff
@@ -1,40 +1,106 @@
 
   fun getMovements(agencyId: String, date: LocalDate): List<Arrival> {
-    val arrivals = basmService.getArrivals(agencyId, date, date).map { addPrisonData(it) } + prisonService.getTransfers(agencyId, date)
+    val arrivals =
+      basmService
+        .getArrivals(agencyId, date, date)
+        .map { augmentWithPrisonData(it) } +
+        prisonService
+          .getTransfers(agencyId, date)
+
     return confirmedArrivalService.extractConfirmedArrivalFromArrivals(agencyId, date, arrivals)
   }
 
-  private fun addPrisonData(arrival: Arrival): Arrival {
+  private fun augmentWithPrisonData(arrival: Arrival): Arrival {
     val candidates = prisonerSearchService.getCandidateMatches(arrival)
 
     val match = candidates.firstOrNull { arrival.isMatch(it) }
 
     return if (match != null)
-      arrival.copy(pncNumber = match.pncNumber, prisonNumber = match.prisonerNumber, isCurrentPrisoner = match.currentPrisoner)
+      arrival.copy(
+        pncNumber = match.pncNumber,
+        prisonNumber = match.prisonerNumber,
+        isCurrentPrisoner = match.isCurrentPrisoner
+      )
 
     // Ignore any provided prison number, but assume PNC is fine
     else
       arrival.copy(prisonNumber = null)
   }
 
+  @PreAuthorize("hasRole('BOOKING_CREATE') and hasAuthority('SCOPE_write')")
+  fun confirmArrival(
+    moveId: String,
+    confirmArrivalDetail: ConfirmArrivalDetail
+  ): ConfirmArrivalResponse {
+
+    val arrival = getMovement(moveId)
+
+    if (arrival.isCurrentPrisoner) throw IllegalArgumentException("The arrival is known to NOMIS and has a current booking. This scenario is not supported.")
+
+    return when (arrival.prisonNumber) {
+      null -> createAndAdmitOffender(confirmArrivalDetail, moveId)
+      else -> admitOffenderOnNewBooking(confirmArrivalDetail, moveId, arrival.prisonNumber)
+    }
+  }
+
+  private fun admitOffenderOnNewBooking(
+    confirmArrivalDetail: ConfirmArrivalDetail,
+    moveId: String,
+    prisonNumber: String
+  ): ConfirmArrivalResponse {
+    prisonService.admitOffenderOnNewBooking(prisonNumber, confirmArrivalDetail)
+    confirmedArrivalService.add(
+      movementId = moveId,
+      prisonNumber = prisonNumber,
+      prisonId = confirmArrivalDetail.prisonId!!,
+      bookingId = 0,
+      arrivalDate = LocalDate.now(),
+      arrivalType = ArrivalType.NEW_BOOKING_EXISTING_OFFENDER
+    )
+    return ConfirmArrivalResponse(offenderNo = prisonNumber)
+  }
+
+  private fun createAndAdmitOffender(
+    confirmArrivalDetail: ConfirmArrivalDetail,
+    moveId: String
+  ): ConfirmArrivalResponse {
+    val prisonNumber = prisonService.createOffender(confirmArrivalDetail)
+    prisonService.admitOffenderOnNewBooking(prisonNumber, confirmArrivalDetail)
+    confirmedArrivalService.add(
+      movementId = moveId,
+      prisonNumber = prisonNumber,
+      prisonId = confirmArrivalDetail.prisonId!!,
+      bookingId = 0,
```

Comment:
@bell-pa is it any way we can set correct booking Id?

Reply:
Done.